File: Grokking/js/lib/quicksort.js
const quicksort = (list) => {
  if(list.length < 2) return list;
  const pivot = list.length / 2 | 0;
  const [ left, right ] = list.reduce(
    ([ left, right ], cur, index) => {
      if(index === pivot) return [ left, right ];
      if(cur < list[pivot]) return [ [ ...left, cur ], right ];
      return [ left, [ ...right, cur ] ];
    },
    [ [], [] ]
  );
  return [ ...(quicksort(left)), list[pivot], ...(quicksort(right)) ];
}

export default quicksort;

File: Grokking/R/quicksort.R
quicksort <- function(vec) {
	if (length(vec) < 1) return(vec)
	pivot <- vec[1]
	left <- vec[vec < pivot]
	right <- vec[vec > pivot]
	c(quicksort(left), vec[vec == pivot], quicksort(right))
}

print(quicksort(c(8, 2, 6, 5, 1, 3, 7)))
print(quicksort(c("Zebra", "Bird", "Panda", "Dog", "Cat")))
